subprojects/base-services/src/main/java/org/gradle/internal/service/DefaultServiceRegistry.java

```diff
@@ -891,19 +891,9 @@ private ConstructorService(DefaultServiceRegistry owner, Class<?> serviceType) {
             if (serviceType.isInterface()) {
                 throw new ServiceValidationException("Cannot register an interface for construction.");
             }
-            Constructor<?>[] constructors = serviceType.getDeclaredConstructors();
-            Constructor<?> match = null;
-            for (Constructor<?> constructor : constructors) {
-                if (Modifier.isPrivate(constructor.getModifiers())) {
-                    continue;
-                }
-                if (match != null) {
-                    throw new ServiceValidationException(String.format("Expected a single non-private constructor for %s.", format(serviceType)));
-                }
-                match = constructor;
-            }
-            if (match == null) {
-                throw new ServiceValidationException(String.format("Expected a single non-private constructor for %s.", format(serviceType)));
```

Comment:
The scope of constructors that can be injected end up being really restricted here.